include/multipass/virtual_machine.h

```diff
@@ -91,22 +92,26 @@ class VirtualMachine : private DisabledCopyMove
     virtual int get_num_snapshots() const noexcept = 0;
     virtual std::shared_ptr<const Snapshot> get_snapshot(const std::string& name) const = 0;
     virtual std::shared_ptr<Snapshot> get_snapshot(const std::string& name) = 0;
-    virtual std::shared_ptr<const Snapshot> take_snapshot(const QDir& snapshot_dir, const VMSpecs& specs,
-                                                          const std::string& name, const std::string& comment) = 0;
-    virtual void delete_snapshot(const QDir& snapshot_dir, const std::string& name) = 0;
-    virtual void restore_snapshot(const QDir& snapshot_dir, const std::string& name, VMSpecs& specs) = 0;
-    virtual void load_snapshots(const QDir& snapshot_dir) = 0;
+    virtual std::shared_ptr<const Snapshot> take_snapshot(const VMSpecs& specs, const std::string& name,
+                                                          const std::string& comment) = 0;
+    virtual void delete_snapshot(const std::string& name) = 0;
+    virtual void restore_snapshot(const std::string& name, VMSpecs& specs) = 0;
+    virtual void load_snapshots() = 0;
 
     VirtualMachine::State state;
     const std::string vm_name;
+    const QDir instance_dir;
     std::condition_variable state_wait;
     std::mutex state_mutex;
     std::optional<IPAddress> management_ip;
     bool shutdown_while_starting{false};
 
 protected:
-    VirtualMachine(VirtualMachine::State state, const std::string& vm_name) : state{state}, vm_name{vm_name} {};
-    VirtualMachine(const std::string& vm_name) : VirtualMachine(State::off, vm_name){};
+    VirtualMachine(VirtualMachine::State state, const std::string& vm_name, const Path& instance_dir)
+        : state{state}, vm_name{vm_name}, instance_dir{QDir{instance_dir}} {};
+    VirtualMachine(const std::string& vm_name, const Path& instance_dir)
+        : VirtualMachine(State::off, vm_name, instance_dir){};
+    VirtualMachine(const std::string& vm_name) : VirtualMachine(vm_name, ""){};
```

Comment:
This is here because of constructors in [`stub_virtual_machine.h`](https://github.com/canonical/multipass/pull/3130/files#diff-9083bcbd10dd40a4be12c75ca0c1035766f76711baa70cc4d4f276f4a1f26756L31-L37). The constructors in [`virtual_machine.h`](https://github.com/canonical/multipass/pull/3130/files#diff-d847b18fbc46d755306da3eb6cac52de71043c1cc87ff345cdf8b034ad141ffaL109-L110) are unchanged apart from the additional parameter for `instance_dir` making it a bit unsightly. Any ideas of how to make this look nicer?

Reply:
The onus of providing a stub directory should be on the stub VM. instead of adding this constructor, the stub VM can call one of the other constructors, passing in a fake `instance_dir`. To be sure that tests don't mess around with disk in unproper places, the best would be to create a temporary directory (e.g. with `QTemporaryDir`).